US87160: Caching-related abstractions for resource/subject attribute readers

anubhavi25 opened the pull request:
… connector

Signed-off-by: Frank <frank.gasparovic@ge.com>

FrankGasparovic commented on acs-integration-tests/pom.xml:
```diff
@@ -496,6 +503,7 @@
                                 <ZAC_CLIENT_SECRET>${ZAC_CLIENT_SECRET}</ZAC_CLIENT_SECRET>
                             </systemPropertyVariables>
                             <excludes>
```
Add ACSMetering until we figure out how to fix that test?

anubhavi25 replied:
Was added as part of DE22023

FrankGasparovic commented on acs-integration-tests/src/test/java/com/ge/predix/integration/test/ExternalResourceAttributeReaderIT.java:
```diff
@@ -0,0 +1,137 @@
+    private static final String ZONE = "test-zone";
+    private static final String RESOURCE_ID = "aircraftPart/13a71359-db68-4602-aac5-a8fa401c3194";
+    private static final String ASSET_ADAPTER_URI = "http://localhost:8080/v1/attribute/";
+    private static final String ASSET_ATTRIBUTES_JSON =
```
Store JSON in src/test/resources like we do for other integration tests

anubhavi25 replied:
Done (however this file has been moved out to another branch)

FrankGasparovic reviewed (commented):
Looks good

FrankGasparovic commented on service/src/test/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationWithAttributeConnectorTest.java:
```diff
@@ -24,8 +24,8 @@
-import com.ge.predix.acs.attribute.connectors.ResourceAttributeConnector;
-import com.ge.predix.acs.attribute.connectors.SubjectAttributeConnector;
+import com.ge.predix.acs.attribute.readers.ExternalResourceAttributeReader;
+import com.ge.predix.acs.attribute.readers.ExternalSubjectAttributeReader;
```
Should we stick to calling everything connectors or readers? We have these ExternalResourceAttributeReader and ExternalSubjectAttributeReader, but the test is called PolicyEvaluationWithAttributeConnectorTest.java, and in the JPA story we created the AttributeConnectorEntity @sanjeevchopra. Just seems like we are making it confusing for no reason.

sanjeevchopra replied:
point taken, @FrankGasparovic lets discuss naming tomorrow.

anubhavi25 replied:
As discussed, we're sticking to the `Reader` suffix for connector-related internal abstractions and the `Connector` suffix for external abstractions

irinaepshteyn commented on acs-integration-tests/pom.xml:
```diff
@@ -621,6 +629,56 @@
 
         <profile>
+            <id>redis</id>
+            <dependencies>
```
Since we added another profile, please update (or add a story to update) https://github.build.ge.com/predix/acs-ci-config/blob/master/AcsTestsReadme.md

irinaepshteyn replied:
In the past it was decided to place test reports outside of target directory so reports are not removed by 'mvn clean' and retained by Jenkins for each profile.  Refer to reportsDirectory and summaryFile in the failsafe plugin configuration for, for example, public profile.
It might not be relevant any longer since each profile is run by the separate Jenkins job now.
We need to decide if we still want to maintain custom output directories and be consistent in every profile.

anubhavi25 replied:
Added changes to `AcsTestsReadme.md` in the master branch of the acs-ci-config Git repo. Also these changes to add the `redis` Maven profile have been pulled out to another branch but when I re-introduce it I'll look into how to standardize the test output directories.

irinaepshteyn commented on acs-integration-tests/pom.xml:
```diff
@@ -991,6 +1051,7 @@
                             <excludes>
+                                <exclude>**/ExternalResourceAttributeReaderIT.java</exclude>
                                 <exclude>**/NonHierarchical*IT.java</exclude>
                             </excludes>
```
Since for acceptance profile we only execute single ACSAcceptanceIT test, excludes section is not needed.  Please remove the whole section.

anubhavi25 replied:
Done

irinaepshteyn commented on service/src/main/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReader.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public Set<Attribute> getAttributesFromAdapter(final String zoneId, final String resourceId) {
+        // TODO: Add an inline call to the Asset Adapter based on configuration
```
Could we change it to private?

anubhavi25 replied:
Made it package-private to allow testing

irinaepshteyn commented on service/src/main/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReader.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public String getZoneId() {
+        return ((ZoneOAuth2Authentication) SecurityContextHolder.getContext().getAuthentication()).getZoneId();
```
Could we change it to private?

anubhavi25 replied:
Made it package-private to allow testing

irinaepshteyn reviewed (changes requested):
These are my initial comments.  I will continue reviewing...

irinaepshteyn commented on service/src/test/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReaderTest.java:
```diff
@@ -0,0 +1,71 @@
+import com.ge.predix.acs.policy.evaluation.cache.InMemoryPolicyEvaluationCache;
+
+public class ExternalResourceAttributeReaderTest {
+
```
This test is written to test mostly InMemoryAttributeCache and not ExternalResourceAttributeReader.  It has to be renamed to com.ge.predix.acs.attribute.cache.InMemoryAttributeCacheTest.
For testing ExternalResourceAttributeReader it should be sufficient to mock cache to return no attributes and mock getAttributesFromAdapter to return a set of attributes and compare what's returned from actual getAttributes.

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/attribute/cache/AbstractAttributeCache.java:
```diff
@@ -0,0 +1,62 @@
+import com.ge.predix.acs.policy.evaluation.cache.AbstractPolicyEvaluationCache;
+
+@Component
+public abstract class AbstractAttributeCache implements AttributeCache {
```
* This class does not need to know the difference between subject of resource. Just (id, attributes)
* Both resource and subject connector will have one instance of each of this cache
* On that note most of the connector implementation will be subject/resource agnostic.

sanjeevchopra replied:
Discussed this with @anubhavi25 .  Added notes in redis caching story. 
* this pull request will be limited to in-memory caching only
* initially a common cache across zones is ok, however we need to limit the storage impact on ACS. Please add 2 env vars - MAX_ATTRIBUTE_CACHE_ENTRIES (default 10K), MAX_ATTRIBUTES_PER_ID  (default 100?)
* we should adjust the mem allocation to ACS accordingly ~ +1gb  (1k per attribute x 100 attributes per user/resource x 10,000 users/resources)
* key should be zone+id.   One cache each for attribute and subject connectors.

anubhavi25 replied:
Done. We can make the relevant config changes to bump up the app memory in acs-ci-config as part of US87168. Also, as discussed, Redis-related changes will be pulled out to the US87168 branch.

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/attribute/cache/AttributeCache.java:
```diff
@@ -0,0 +1,21 @@
+
+import com.ge.predix.acs.model.Attribute;
+
+public interface AttributeCache {
```
Dont see much value in this interface. Just make this the abstract class?

anubhavi25 replied:
Removed in lieu of `AbstractAttributeCache`

irinaepshteyn commented on service/src/test/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReaderTest.java:
```diff
@@ -0,0 +1,71 @@
+        Set<Attribute> cachedAttributes = this.externalResourceAttributeReader.getAttributes(RESOURCE_ID);
+
+        for (Attribute cachedAttribute : cachedAttributes) {
+            Assert.assertTrue(attributes.contains(cachedAttribute));
```
Instead of 'for' loop could use Assert.assertTrue(attributes.containsAll(cachedAttributes)) or Assert.assertEquals(attributes, cachedAttributes);

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/attribute/cache/AbstractAttributeCache.java:
```diff
@@ -0,0 +1,62 @@
+
+    public abstract long getExpiration(String key);
+
+    public abstract long getConfiguredTtl();
```
please add time unit to the method name

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/attribute/cache/AbstractAttributeCache.java:
```diff
@@ -0,0 +1,62 @@
+
+    public abstract Set<Attribute> get(String key);
+
+    public abstract long getExpiration(String key);
```
please add time unit to the method name

anubhavi25 replied:
Done (renamed this to `getRemainingTtlInSeconds`)

sanjeevchopra commented on service/pom.xml:
```diff
@@ -335,6 +335,11 @@
             <artifactId>jedis</artifactId>
             <version>${jedis.version}</version><!--$NO-MVN-MAN-VER$-->
         </dependency>
+        <dependency>
```
Please check with Anurag on process to get new dependencies approved.

anubhavi25 replied:
Now using EhCache which, according to Anurag, is under Apache licensing

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReader.java:
```diff
@@ -0,0 +1,46 @@
+import com.ge.predix.uaa.token.lib.ZoneOAuth2Authentication;
+
+@Component
+public class ExternalResourceAttributeReader implements ResourceAttributeReader {
```
`ExternalAttributeSourceReader` seems more accurate. I still like `ResourceAttributeConnector` , though :)

anubhavi25 replied:
As discussed, sticking to `ExternalResourceAttributeReader`

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReader.java:
```diff
@@ -0,0 +1,46 @@
+        return attributes;
+    }
+
+    public Set<Attribute> getAttributesFromAdapter(final String zoneId, final String resourceId) {
```
getAttributesFromAdapters  (plural)

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java:
```diff
@@ -37,6 +27,17 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import org.codehaus.jackson.map.ObjectMapper;
```
pls revert changes to this package

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/test/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReaderTest.java:
```diff
@@ -0,0 +1,71 @@
+        Mockito.doReturn(ZONE).when(this.externalResourceAttributeReader).getZoneId();
+        Set<Attribute> cachedAttributes = this.externalResourceAttributeReader.getAttributes(RESOURCE_ID);
+
+        for (Attribute cachedAttribute : cachedAttributes) {
```
use equals()

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/test/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReaderTest.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Test
+    public void testGetAttributes() throws Exception {
+        Set<Attribute> attributes = Collections.singleton(
```
expectedAttributes

anubhavi25 replied:
Done

sanjeevchopra commented on service/src/test/java/com/ge/predix/acs/attribute/readers/ExternalResourceAttributeReaderTest.java:
```diff
@@ -0,0 +1,71 @@
+        Mockito.doReturn(attributes).when(this.externalResourceAttributeReader)
+                .getAttributesFromAdapter(ZONE, RESOURCE_ID);
+        Mockito.doReturn(ZONE).when(this.externalResourceAttributeReader).getZoneId();
+        Set<Attribute> cachedAttributes = this.externalResourceAttributeReader.getAttributes(RESOURCE_ID);
```
actualAttributes

anubhavi25 replied:
Done

irinaepshteyn commented on acs-integration-tests/src/test/java/com/ge/predix/integration/test/ExternalResourceAttributeReaderIT.java:
```diff
@@ -0,0 +1,137 @@
+        Long ttl = this.redisTemplate.getExpire(resourceAttributesKey, TimeUnit.SECONDS);
+        Assert.assertTrue(ttl > 0L && ttl <= this.attributeCache.getConfiguredTtl());
+    }
+}
```
We need to add another IT: PolicyEvalCachingWithExternalResourceAttributeReaderIT to assert that policy decision is reevaluated if attributes cache is expired and attributes are refreshed.

anubhavi25 replied:
This will be added as part of US87156

irinaepshteyn reviewed (dismissed):
Additional review comments.

anubhavi25 commented:
@sanjeevchopra @irinaepshteyn please review again as there were non-trivial changes in the last commit (i.e. removed Redis-related changes, moved to EhCache and incorporated cache-related constraints)

Commit:
US87160: Caching-related abstractions for resource/subject attribute readers

Signed-off-by: Sanjeev <chopra@ge.com>
Signed-off-by: Frank <frank.gasparovic@ge.com>
